command/root.go

```diff
@@ -401,14 +403,44 @@ func ExpandAlias(args []string) ([]string, error) {
 			return empty, fmt.Errorf("not enough arguments for alias: %s", expansion)
 		}
 
-		newArgs, err := shlex.Split(expansion)
-		if err != nil {
-			return nil, err
+		isExternal := strings.HasPrefix(expansion, "!")
+
+		if !isExternal {
+
+			newArgs, err := shlex.Split(expansion)
+			if err != nil {
+				return nil, err
+			}
+
+			return append(newArgs, extraArgs...), nil
+		}
+
+		defaultShell := "sh"
+		if os.Getenv("SHELL") != "" {
+			defaultShell = os.Getenv("SHELL")
+		}
+
+		for _, arg := range extraArgs {
+			if !strings.HasPrefix(arg, "-") {
+				expansion += fmt.Sprintf(" %q ", arg)
+			} else {
+				expansion += fmt.Sprintf(" %s ", arg)
```

Comment:
I have some worries about how this appends additional arguments. First of all, arguments are appended as strings to `expansion`, which is risky because we then need to ensure that they are quoted or otherwise escaped. Second, I don't think it gains us anything to differentiate between arguments that start with `-` vs. those that don't, since both can contain spaces and other special characters.

Using `%q` to quote a string works only for the simplest values, but we should keep in mind that this produces Go syntax which is not necessarily compatible with shells. For instance, newlines and tab characters will be encoded as `\n` and `\t`, which in bash will result in literal `\n` and `\t` strings. Furthermore, using `%q` doesn't escape characters that may have special meaning in the shell; most notably `$`.

The safest way to pass additional arguments would be to avoid encoding them as escaped strings and instead passing them as extra arguments in `exec.Command`. The shell equivalent of that would be:
```
$ sh -c 'echo 1:$1 2:$2' -- foo bar
1:foo 2:bar

$ bash -c 'echo 1:$1 2:$2' -- foo bar
1:foo 2:bar

$ zsh -c 'echo 1:$1 2:$2' -- foo bar
1:foo 2:bar
```

The alias would then have to explicitly choose where to place positional parameters `$1`, `$2`, `"$@"`, etc.

Unfortunately, extra arguments appended after `-c <COMMAND>` are downright ignored by fish https://github.com/fish-shell/fish-shell/issues/2314

So for fish we would have to do something different or, if we go back to square one, we'd have to ensure that all arguments appended to `expansion` would have to be shell-escaped so that no characters within have any special meaning.

Reply:
I've found how git handles it: basically it always does `sh -c '<command> "$@"' <args>` regardless of the SHELL environment variable. https://github.com/git/git/blob/050319c2ae82f2fac00eac9d80a1d91394d99aec/run-command.c#L266-L291

I think that we could follow the same principle so that people's aliases work identically regardless of their current interactive shell?

Reply:
I'm fine with just using `sh`; I considered expanding `$SHELL` in case people wanted to rely on their shell-specific aliases. I feel like we can worry about that if it's requested.

As for the additional argument handling, I'm very on board. I hated the code I wrote and knew several ways it would break; i just wasn't sure how to improve on it. Mandating `sh` and using the `--` trick (TIL) is great. I'll work on that now.